article/app/topmentions/TopMentionsS3Client.scala

```diff
@@ -0,0 +1,97 @@
+import com.amazonaws.services.s3.model.{GetObjectRequest, S3Object}
+import com.amazonaws.util.IOUtils
+import common.GuLogging
+import conf.Configuration
+import play.api.libs.json.{JsError, JsSuccess, Json}
+import services.S3
+import topmentions.S3ObjectImplicits.RichS3Object
+import topmentions.TopMentionsResponse._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+trait TopMentionsS3Client {
+  def getListOfKeys(): Future[List[String]]
+  def getObject(key: String): Future[TopMentionsDetails]
+}
+
+final class TopMentionsS3ClientImpl extends TopMentionsS3Client with S3 with GuLogging {
+  lazy val optionalBucket: Option[String] = Configuration.aws.topMentionsStoreBucket
+
+  def getListOfKeys(): Future[List[String]] = {
+    getClient { client =>
+      Try {
+        val s3ObjectList = client.listObjects(getBucket).getObjectSummaries().asScala.toList
+        s3ObjectList.map(_.getKey)
+      } match {
+        case Success(value) =>
+          log.info(s"got list of ${value.length} top mentions from S3")
+          Future.successful(value)
+        case Failure(exception) =>
+          log.error(s"failed in getting the list of top mentions from S3 - ${exception.getMessage}")
+          Future.failed(exception)
+      }
+    }
+  }
+
+  def getObject(key: String): Future[TopMentionsDetails] = {
+    getClient { client =>
+      Try {
+        val request = new GetObjectRequest(getBucket, key)
+        client.getObject(request).parseToTopMentionsDetails
+      } match {
+        case Success(value) =>
+          log.info(s"got topMentionResponse from S3 for key ${key}")
+          Future.successful(value)
+        case Failure(exception) =>
+          log.error(s"S3 retrieval failed for key ${key} - ${exception.getMessage}")
+          Future.failed(exception)
+      }
+    }
+  }
+
+  private def getBucket() = {
+    optionalBucket.getOrElse(
+      throw new RuntimeException("top mention bucket config is empty, make sure config parameter has value"),
+    )
+  }
+
+  private def getClient[T](callS3: AmazonS3 => Future[T]) = {
+    client
+      .map { callS3(_) }
+      .getOrElse(Future.failed(new RuntimeException("No client exist for TopMentionsS3Client")))
+  }
+}
+
+object S3ObjectImplicits {
+  implicit class RichS3Object(s3Object: S3Object) extends GuLogging {
+    def parseToTopMentionsDetails: TopMentionsDetails = {
+      val json = Json.parse(asString(s3Object))
+
+      Json.fromJson[TopMentionsDetails](json) match {
+        case JsSuccess(topMentionResponse, __) =>
+          log.debug(s"Parsed topMentionResponse from S3 for key ${s3Object.getKey}")
+          topMentionResponse
+        case JsError(errors) =>
+          val errorPaths = errors.map { error => error._1.toString() }.mkString(",")
+          log.error(s"Error parsing topMentionResponse from S3 for key ${s3Object.getKey} paths: ${errorPaths}")
+          throw TopMentionJsonParseException(
```

Comment:
We usually try to avoid throwing errors in frontend and use Futures whenever we can

Reply:
Hi @AshCorr thanks for looking. This `parseToTopMentionsDetails` method is used within a Try that's supposed to handle any error regarding the S3 call or parsing. 

I see what your point is about trying to avoid errors. I wonder if Future is best fit for this case as this parsing doesn't need to be done concurrently? 

Maybe I can use an Option or Either? But doing that, we would need to add a case in the caller as followed: 
```
        case Success(Some(value)) =>
          log.info(s"got list of ${value.length} top mentions from S3")
          Future.successful(value)
        case Success(None) =>
          log.info(s"could not parse top mentions from S3")
          Future.failed(throw TopMentionJsonParseException("..."))
        case Failure(exception) =>
          log.error(s"failed in getting the list of top mentions from S3 - ${exception.getMessage}")
          Future.failed(exception)
```

What do you think?

Reply:
Agree would be nice to avoid the throw, Try is good and returning an Option is another option!

Reply:
Thank @AshCorr it looks great. I didn't need to use the map at the end and instead used a flatten

```
      Try {
        val request = new GetObjectRequest(getBucket, key)
        client.getObject(request).parseToTopMentionsDetails
      }.flatten match {
        case Success(value) =>
          log.info(s"got topMentionResponse from S3 for key ${key}")
          Future.successful(value)
        case Failure(exception) =>
          log.error(s"S3 retrieval failed for key ${key}", exception)
          Future.failed(exception)
      }
```

because `client.getObject(request)` is not returning a future and is returning an `S3Object`. The parseToTopMentionsDetails was added as an extension method for S3Object.

Reply:
Thanks @alinaboghiu yes `Try` seemed to be good for this case